File: java/src/main/java/io/github/athxx/ason/Value.java
package io.github.athxx.ason;

import java.util.*;

/**
 * Represents an ASON value.
 */
public class Value {
    
    public enum Type {
        NULL, BOOL, INTEGER, FLOAT, STRING, ARRAY, OBJECT
    }
    
    private final Type type;
    private final Object data;
    private final List<String> keys; // for preserving object key order
    
    private Value(Type type, Object data) {
        this(type, data, null);
    }
    
    private Value(Type type, Object data, List<String> keys) {
        this.type = type;
        this.data = data;
        this.keys = keys;
    }
    
    // Factory methods
    public static Value ofNull() {
        return new Value(Type.NULL, null);
    }
    
    public static Value ofBool(boolean v) {
        return new Value(Type.BOOL, v);
    }
    
    public static Value ofInteger(long v) {
        return new Value(Type.INTEGER, v);
    }
    
    public static Value ofFloat(double v) {
        return new Value(Type.FLOAT, v);
    }
    
    public static Value ofString(String v) {
        return new Value(Type.STRING, v);
    }
    
    public static Value ofArray() {
        return new Value(Type.ARRAY, new ArrayList<Value>());
    }
    
    public static Value ofArray(Value... items) {
        List<Value> list = new ArrayList<>(Arrays.asList(items));
        return new Value(Type.ARRAY, list);
    }
    
    public static Value ofObject() {
        return new Value(Type.OBJECT, new LinkedHashMap<String, Value>(), new ArrayList<>());
    }
    
    // Type checks
    public Type getType() { return type; }
    public boolean isNull() { return type == Type.NULL; }
    public boolean isBool() { return type == Type.BOOL; }
    public boolean isInteger() { return type == Type.INTEGER; }
    public boolean isFloat() { return type == Type.FLOAT; }
    public boolean isNumber() { return isInteger() || isFloat(); }
    public boolean isString() { return type == Type.STRING; }
    public boolean isArray() { return type == Type.ARRAY; }
    public boolean isObject() { return type == Type.OBJECT; }
    
    // Value getters
    public boolean asBool() {
        return type == Type.BOOL ? (Boolean) data : false;
    }
    
    public long asInteger() {
        if (type == Type.INTEGER) return (Long) data;
        if (type == Type.FLOAT) return ((Double) data).longValue();
        return 0;
    }
    
    public double asFloat() {
        if (type == Type.FLOAT) return (Double) data;
        if (type == Type.INTEGER) return ((Long) data).doubleValue();
        return 0;
    }
    
    public String asString() {
        return type == Type.STRING ? (String) data : "";
    }
    
    // Array operations
    @SuppressWarnings("unchecked")
    public int size() {
        if (type == Type.ARRAY) return ((List<Value>) data).size();
        if (type == Type.OBJECT) return ((Map<String, Value>) data).size();
        return 0;
    }
    
    @SuppressWarnings("unchecked")
    public Value get(int index) {
        if (type != Type.ARRAY) return null;
        List<Value> list = (List<Value>) data;
        if (index < 0 || index >= list.size()) return null;
        return list.get(index);
    }
    
    @SuppressWarnings("unchecked")
    public void push(Value item) {
        if (type != Type.ARRAY) return;
        ((List<Value>) data).add(item);
    }
    
    @SuppressWarnings("unchecked")
    public List<Value> items() {
        if (type != Type.ARRAY) return Collections.emptyList();
        return Collections.unmodifiableList((List<Value>) data);
    }
    
    // Object operations
    @SuppressWarnings("unchecked")
    public Value get(String key) {
        if (type != Type.OBJECT) return null;
        return ((Map<String, Value>) data).get(key);
    }
    
    @SuppressWarnings("unchecked")
    public void set(String key, Value value) {
        if (type != Type.OBJECT) return;
        Map<String, Value> map = (Map<String, Value>) data;
        if (!map.containsKey(key)) {
            keys.add(key);
        }
        map.put(key, value);
    }
    
    public List<String> keys() {
        if (type != Type.OBJECT) return Collections.emptyList();
        return Collections.unmodifiableList(keys);
    }
    
    @Override
    public String toString() {
        return "Value{type=" + type + ", data=" + data + "}";
    }
}

File: kotlin/src/main/kotlin/io/github/athxx/ason/Value.kt
package io.github.athxx.ason

/**
 * Represents an ASON value.
 */
sealed class Value {
    
    object Null : Value() {
        override fun toString() = "null"
    }
    
    data class Bool(val value: Boolean) : Value() {
        override fun toString() = if (value) "true" else "false"
    }
    
    data class Integer(val value: Long) : Value() {
        override fun toString() = value.toString()
    }
    
    data class Float(val value: Double) : Value() {
        override fun toString() = value.toString()
    }
    
    data class Str(val value: String) : Value() {
        override fun toString() = "\"$value\""
    }
    
    data class Array(val items: MutableList<Value> = mutableListOf()) : Value() {
        operator fun get(index: Int): Value? = items.getOrNull(index)
        fun add(item: Value) = items.add(item)
        val size: Int get() = items.size
    }
    
    data class Object(
        private val map: MutableMap<String, Value> = mutableMapOf(),
        val keys: MutableList<String> = mutableListOf()
    ) : Value() {
        operator fun get(key: String): Value? = map[key]
        
        operator fun set(key: String, value: Value) {
            if (!map.containsKey(key)) {
                keys.add(key)
            }
            map[key] = value
        }
        
        val size: Int get() = map.size
        
        fun entries(): List<Pair<String, Value>> = keys.map { it to map[it]!! }
    }
    
    // Type checks
    val isNull: Boolean get() = this is Null
    val isBool: Boolean get() = this is Bool
    val isInteger: Boolean get() = this is Integer
    val isFloat: Boolean get() = this is Float
    val isNumber: Boolean get() = isInteger || isFloat
    val isString: Boolean get() = this is Str
    val isArray: Boolean get() = this is Array
    val isObject: Boolean get() = this is Object
    
    // Value getters
    fun asBool(): Boolean = (this as? Bool)?.value ?: false
    fun asInteger(): Long = when (this) {
        is Integer -> value
        is Float -> value.toLong()
        else -> 0
    }
    fun asFloat(): Double = when (this) {
        is Float -> value
        is Integer -> value.toDouble()
        else -> 0.0
    }
    fun asString(): String = (this as? Str)?.value ?: ""
    fun asArray(): Array? = this as? Array
    fun asObject(): Object? = this as? Object
    
    companion object {
        fun ofNull() = Null
        fun ofBool(v: Boolean) = Bool(v)
        fun ofInteger(v: Long) = Integer(v)
        fun ofFloat(v: Double) = Float(v)
        fun ofString(v: String) = Str(v)
        fun ofArray(vararg items: Value) = Array(items.toMutableList())
        fun ofObject() = Object()
    }
}

